### PR TITLE
Feat/firestore listeners and filters

### LINKED ISSUE #274: Improve event repository for perf and integrity
## Description
The `EventRepositoryFirestore` class exhibits performance bottlenecks in event fetching and filtering, particularly with large event lists and client-side filtering, and risks data inconsistencies due to non-atomic updates in `addEvent`, `editEvent`, and `deleteEvent`. Improvements include using Firestore batched writes or transactions for atomic operations, optimizing event queries with server-side filtering where possible, caching frequently accessed events, and enhancing error handling and logging for diagnostics.

## Acceptance Criteria

* **Performance**:
  - Event fetching in `getFilteredEvents` and `fetchEventsByIds` is optimized by moving price and location filters to server-side Firestore queries where feasible, reducing client-side processing.
  - Queries for large event ID lists (e.g., in `getUserEvents`) are paginated or cached locally to minimize Firestore round-trips, with response times under 500ms for up to 100 events.
  - Real-time updates in `listenToFilteredEvents` are throttled or debounced to prevent UI overload during frequent Firestore snapshots.

* **Integrity**:
  - `addEvent`, `editEvent`, and `deleteEvent` use Firestore batched writes or transactions to ensure atomic updates across event documents and user arrays (`ownedEventIds`, `joinedEventIds`, `savedEventIds`), preventing partial updates.
  - Participant IDs in `editEvent` are validated to exist in the `users` collection before updates, with invalid IDs logged and skipped.
  - Failed Firestore operations (e.g., network errors) are logged with specific event/user IDs and skipped without crashing, ensuring repository stability.

* **Error Handling**:
  - Invalid event data (e.g., missing `uid`, invalid `date`) is rejected with specific exceptions (e.g., `IllegalArgumentException`) and logged with context (e.g., event ID, operation type).
  - Tag truncation in `getFilteredEvents` and `listenToFilteredEvents` (due to `FIRESTORE_QUERY_LIMIT`) throws a custom exception or returns a partial result warning, logged for debugging.
  - Authentication errors (e.g., null user ID in `friendsOnly` filtering) throw an `IllegalStateException` with a clear message.

* **UI Preservation**:
  - Existing event retrieval, filtering, and real-time update behaviors remain unchanged from the caller's perspective, ensuring no disruption to UI components relying on `EventRepository`.

### PR DESCRIPTION
**Pull request by e-bernier**
## Description
This PR significantly improves the performance and correctness of event filtering while adding proper real-time listeners for a user’s saved and joined events. This is the final PR about EventRepositoryFirestore refactor focused on performance and integrity of the data.

Key improvements:
- Smart query strategies that push as much filtering as possible to Firestore (dates, price, tags, friends-only) while correctly applying client-side filters for geo-distance and popularOnly
- Proper handling of `added/modified/removed` events in `listenToFilteredEvents`
- New `listenToSavedEvents` and `listenToJoinedEvents` methods on `EventRepository` that react instantly when a saved/joined event is modified or deleted
- Robust dynamic listener management: subscribes/unsubscribes to individual event documents as the user’s saved/joined lists change

These changes eliminate stale data in the “Saved” and “Joined” tabs, drastically reduce over-fetching on the map screen, and prepare the repository for efficient pagination in future PRs.

**Related Issue:** Closes #274 

---
## Changes

**Implementation:**
- Add `listenToSavedEvents` and `listenToJoinedEvents` to `EventRepository` interface
- Refactor `getFilteredEvents` and `listenToFilteredEvents` in `EventRepositoryFirestore` with query strategy selection (FRIENDS → TAGS → BASIC)
- Implement chunked `whereIn` queries for friends-only filtering
- Apply client-side geo-distance, popularOnly (>30 participants), and tag filtering when required
- Implement generic `listenToUserEvents` with dynamic per-event listeners that detect modifications and deletions
- Add `POPULAR_EVENT = 30` threshold and helper utilities

**Tests:**
- Added comprehensive unit tests for all query strategies and filter combinations
- Full test coverage of `listenToFilteredEvents` with client-side filtering
- Extensive tests for `listenToSavedEvents` and `listenToJoinedEvents`: addition/removal, modification, deletion, error cases, dynamic listener lifecycle, and cleanup

---
## Testing
- [x] Unit tests pass (100+ new assertions)
- [x] Instrumentation tests pass
- [x] Manual testing completed (map filters, saved/joined tabs update instantly on event changes/deletions)
- [x] Coverage maintained (≥80% on repository layer)

**Test summary:** Added ~450 lines of focused repository tests covering every new code path, edge cases (large participant lists, document deletion, listener errors), and real-time behavior.

---
## Screenshots/Videos

---
## Checklist
- [x] Sufficient documentation and minimal inline comments
- [x] All tests passing
- [x] No new warnings
- [x] If related issue exists: issue has all labels, fields, and milestone filled

### CODE REVIEW

File: app/src/main/java/com/swent/mapin/model/event/EventRepository.kt [hunk 1]
```diff
@@ -107,4 +107,30 @@ interface EventRepository {
...
+   * fetches corresponding events.
+   *
+   * @param userId The unique identifier of the user.
+   * @param onUpdate Callback with added, modified, and removed events.
+   * @return ListenerRegistration to manage the listener.
+   */
+  fun listenToJoinedEvents(
+      userId: String,
+      onUpdate: (List<Event>, List<Event>, List<Event>) -> Unit
+  ): ListenerRegistration
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/model/event/EventRepository.kt]: Change done, i added some more Kdoc and noticed a bug in my code at the same time so thanks :) (with the detection of initalCall when events are added)

File: app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt [hunk 2]
```diff
@@ -350,39 +509,172 @@ class EventRepositoryFirestore(
...
+                  } else {
+                    // Event was modified
+                    val modifiedEvent = eventSnapshot.toEvent()
+                    if (modifiedEvent != null) {
+                      onUpdate(emptyList(), listOf(modifiedEvent), emptyList())
+                    }
+                  }
+                }
+            eventListeners[eventId] = listener
+          }
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt]: This is a problem i also had identified but I think it's outside the scope of the course to do better as we will only have a few users.

File: app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt [hunk 3]
```diff
@@ -350,39 +509,172 @@ class EventRepositoryFirestore(
...
+                    val deletedEvent =
+                        Event(
+                            uid = eventId,
+                            title = "",
+                            description = "",
+                            date = Timestamp.now(),
+                            location = Location("", 0.0, 0.0),
+                            ownerId = "",
+                            tags = emptyList())
+                    onUpdate(emptyList(), emptyList(), listOf(deletedEvent))
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt]: I changed it so that it simply returns the eventId for deletedEvents instead of creating an empty event.

File: app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt [hunk 4]
```diff
@@ -279,24 +283,175 @@ class EventRepositoryFirestore(
...
+    val startTimestamp =
+        Timestamp(filters.startDate.atStartOfDay(ZoneId.systemDefault()).toEpochSecond(), 0)
+    query = query.whereGreaterThanOrEqualTo("date", startTimestamp)
+
+    // Apply endDate filter if provided
+    if (filters.endDate != null) {
+      val endTimestamp =
+          Timestamp(
+              filters.endDate.atTime(23, 59, 59).atZone(ZoneId.systemDefault()).toEpochSecond(), 0)
+      query = query.whereLessThanOrEqualTo("date", endTimestamp)
```
[the remark this answers is not in the document]
Comment (e-bernier) [reply app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt]: Change done, i actually noticed there was already a time Utils doing this so i just use it now.

**Review by naderbahri — APPROVED** [review]
This now looks great to me, it can be merged to main.

File: app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (naderbahri) [reply app/src/main/java/com/swent/mapin/model/event/EventRepositoryFirestore.kt]: That's true, for the moment it's not really important
